datalad/utils.py

```diff
@@ -1216,25 +1216,99 @@ def updated(d, update):
     return d
 
 
+_pwd_mode = None
+
+
+def _switch_to_getcwd(msg, *args):
+    global _pwd_mode
+    _pwd_mode = 'cwd'
+    lgr.warning(
+        msg + ". From now on will be returning os.getcwd(). Directory"
+               " symlinks in the paths will be resolved",
+        *args
+    )
+    # TODO:  we might want to mitigate by going through all flywheighted
+    # repos and tuning up their .paths to be resolved?
+
+
 def getpwd():
     """Try to return a CWD without dereferencing possible symlinks
 
-    If no PWD found in the env, output of getcwd() is returned
+    This function will try to use PWD environment variable to provide a current
+    working directory, possibly with some directories along the path being
+    symlinks to other directories.  Unfortunately, PWD is used/set only by the
+    shell and such functions as `os.chdir` and `os.getcwd` nohow use or modify
+    it, thus `os.getcwd()` returns path with links dereferenced.
+
+    While returning current working directory based on PWD env variable we
+    verify that the directory is the same as `os.getcwd()` after resolving all
+    symlinks.  If that verification fails, we fall back to always use
+    `os.getcwd()`.
+
+    Initial decision to either use PWD env variable or os.getcwd() is done upon
+    the first call of this function.
```

Comment:
The workaround is ugly, but this docstring does a nice job of explaining why it exists.